### PR TITLE
Transformations: Support a rows mode in labels to fields

### PR DESCRIPTION
**Pull request by ryantxu**
The labels to fields mode gets pretty awkward when trying to show a table of key value pairs from labels.  Currently the best we can do is:
1. Labels to fields
2. Filter by name
3. very wide table :(

Ideally we could add a rotate option on the table panel, but that is a bigger ball of wax.  In this PR, we simplify the process of making a clean table from labels by adding a new mode and incorporating a filter option directly on the labels.

Peek 2021-10-27 23-15

TODO:
- [x] docs update

### CODE REVIEW

**Review by mckn — COMMENTED** [review]
The code and tests looks really good! Great work!

**Review by kaydelaney — COMMENTED** [review]
Looks good and works well. 👍 
Just left some small comments

**Comment by ryantxu** [comment]
> One thought I had was we have the "series to rows" transform, wonder if that is a better fit for the mode of the row mode?

I don't think so given that the case there is dealing with the actual values and labels may or may not exist.  I stuck it on this existing transform because it is already clear they are working with labels.  This also adds the feature of filtering the labels directly rather than requiring an additional step.

File: public/app/core/components/TransformersUI/LabelsToFieldsTransformerEditor.tsx [hunk 1]
```diff
@@ -33,25 +42,74 @@ export const LabelsAsFieldsTransformerEditor: React.FC<TransformerUIProps<Labels
     }
   }
 
+  const selected = new Set(options.keepLabels?.length ? options.keepLabels : Object.keys(uniqueLabels));
```
[the remark this answers is not in the document]
Comment (ryantxu) [reply public/app/core/components/TransformersUI/LabelsToFieldsTransformerEditor.tsx]: :+1:  added a `useMemo()`

**Comment by torkelo** [comment]
Without the integrated merge function, the labels to rows transform is now a lot harder to use with prometheus/graphite as they always returns a separate frame per series. You always only see a single result when using this transform. and it's not clear what to do in order to get a single table result. Wonder if we can make table panel suggest a merge transform when the transform tab is open and there are multiple frames.

File: packages/grafana-data/src/transformations/transformers/labelsToFields.ts [hunk 2]
```diff
@@ -77,3 +96,38 @@ export const labelsToFieldsTransformer: SynchronousDataTransformerInfo<LabelsToF
...
+            vals.push(val);
+          }
+        }
+        if (vals.length) {
+          result.push({
+            ...frame,
+            name: getFieldDisplayName(field, frame, data),
+            fields: [
+              { name: 'label', type: FieldType.string, config: {}, values: new ArrayVector(keys) },
+              { name: 'value', type: FieldType.string, config: {}, values: new ArrayVector(vals) },
```
Comment (torkelo) [inline packages/grafana-data/src/transformations/transformers/labelsToFields.ts:124]: Should we capitalize these (Label vs label) , Value vs value? not sure :shrug:  , but I think time series field names are Time and Value (ie capitalized)

**Review by torkelo — APPROVED** [review]
Looks good

**Review by mckn — APPROVED** [review]
LGTM! Great work adding documentation for this 🙌🏻

File: docs/sources/panels/transformations/types-options.md [hunk 3]
```diff
@@ -192,20 +192,33 @@ In the example below, I added two fields together and named them Sum.
 
 ## Labels to fields
 
-This transformation changes time series results that include labels or tags into to a table structure where each label becomes its own field.
+This transformation changes time series results that include labels or tags into to a table structure where each label keys and values
+are included in the table result. The labels can be displayed either as columns or as row values
```
Comment (achatterjee-grafana) [inline docs/sources/panels/transformations/types-options.md:196]:
```suggestion
are included in the table result. The labels can be displayed either as columns or as row values.
```

File: docs/sources/panels/transformations/types-options.md [hunk 4]
```diff
@@ -192,20 +192,33 @@ In the example below, I added two fields together and named them Sum.
...
 
 Given a query result of two time series:
 
 - Series 1: labels Server=Server A, Datacenter=EU
 - Series 2: labels Server=Server B, Datacenter=EU
 
-This would result in a table like this:
+In "Columns" mode, the result will look like this:
```
Comment (achatterjee-grafana) [inline docs/sources/panels/transformations/types-options.md:203]:
```suggestion
In "Columns" mode, the result looks like this:
```

File: docs/sources/panels/transformations/types-options.md [hunk 5]
```diff
@@ -192,20 +192,33 @@ In the example below, I added two fields together and named them Sum.
...
 
 | Time                | Server   | Datacenter | Value |
 | ------------------- | -------- | ---------- | ----- |
 | 2020-07-07 11:34:20 | Server A | EU         | 1     |
 | 2020-07-07 11:34:20 | Server B | EU         | 2     |
 
+In "Rows" mode, the result will contain a table for each series and show each label value like this:
```
Comment (achatterjee-grafana) [inline docs/sources/panels/transformations/types-options.md:210]:
```suggestion
In "Rows" mode, the result has a table for each series and show each label value like this:
```

**Review by achatterjee-grafana — COMMENTED** [review]
Added doc review.

File: packages/grafana-data/src/transformations/transformers/labelsToFields.ts (hunk 2 above)
Comment (ryantxu) [reply packages/grafana-data/src/transformations/transformers/labelsToFields.ts:124]: when capitalized special magic name handling takes over :(  

it ends up with the frame name in the display unless we add additional logic :shrug:

**Pull request merged**